palace/utils/errorindicators.hpp

```diff
@@ -0,0 +1,87 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef PALACE_UTILS_ERROR_INDICATORS_HPP
+#define PALACE_UTILS_ERROR_INDICATORS_HPP
+
+#include "linalg/vector.hpp"
+#include "utils/communication.hpp"
+
+namespace palace
+{
+// Storage for error estimation results from the solve. Required in the AMR loop. This is
+// richer than the IndicatorsAndNormalization because it stores derived quantities, and a
+// communicator for use in adaptation.
+class ErrorIndicators
+{
+public:
+  ErrorIndicators(Vector local, double normalization)
+    : local(std::move(local)), normalization(normalization), n(1)
+  {
+  }
+
+  ErrorIndicators() = default;
+  ErrorIndicators(const ErrorIndicators &) = default;
+  ErrorIndicators(ErrorIndicators &&) = default;
+  ErrorIndicators &operator=(const ErrorIndicators &) = default;
+  ErrorIndicators &operator=(ErrorIndicators &&) = default;
+  ~ErrorIndicators() = default;
+
+  // Return the average normalized local error indicators.
+  const auto &GetLocalErrorIndicators() const { return local; }
+  // Return the global error indicator.
+  inline auto GetGlobalErrorIndicator(MPI_Comm comm) const
+  {
+    constexpr int p = 2;
+    double global_error_indicator =
+        std::transform_reduce(local.begin(), local.end(), 0.0, std::plus(),
+                              [](auto val) { return std::pow(val, p); });
+    Mpi::GlobalSum(1, &global_error_indicator, comm);
+    return std::pow(global_error_indicator, 1.0 / p);
+  }
+  // Return the largest normalized local error indicator.
+  inline auto GetMaxErrorIndicator(MPI_Comm comm) const
+  {
+    double max = local.Max();
+    Mpi::GlobalMax(1, &max, comm);
+    return max;
+  }
+  // Return the smallest normalized local error indicator.
+  inline auto GetMinErrorIndicator(MPI_Comm comm) const
+  {
+    double min = local.Min();
+    Mpi::GlobalMin(1, &min, comm);
+    return min;
+  }
+  // Return the mean normalized local error indicator.
+  inline auto GetMeanErrorIndicator(MPI_Comm comm) const
+  {
+    int size = local.Size();
+    auto global_error_indicator = GetGlobalErrorIndicator(comm);
+    return global_error_indicator / size;
+  }
+  // Return the normalization constant for the absolute error.
+  inline auto GetNormalization() const { return normalization; }
+  // Add a set of indicators to the running totals.
+  void AddIndicators(const ErrorIndicators &indicators);
+  // Reset a running total of error indicators ready for computing a new running average.
+  inline void Reset()
+  {
+    n = 0;
+    local = Vector();
+    normalization = 0;
+  }
+
+protected:
+  // Elemental localized error indicators. Used for marking elements for
+  // refinement and coarsening.
+  Vector local;
```

Comment:
Also, another comment that will be helpful. If you construct these `Vector`s as `ParGridFunction`s over an order-0 L2 FE space, you won't need to add the stuff inside of `PostOperator`. This FE space is best constructed by `SpaceOperator` and the like, probably. It can actually be a standard `GridFunction`, since we never need the neighbor data.

Reply:
Lastly, I'd move this into `fem/`. Seems more appropriate there since it isn't really just a "utility" class anymore.

Reply:
The estimator was originally dumb in the first round, just returning the vector and normalization, but I rolled that into `ErrorIndicator` based on your comments about combining it all together. The downside of `ErrorIndicators` storing the vectors internally is now all of that information is going to leak out into the SEMR loop, which was previously just ingesting the single reduced quantity, hence the return type of the `Solve`. Having rolled all the classes in `errorindicators.hpp` into one, I now think that's a much better interface, as it can be used as just the dumb output of error estimator, as it currently is, but is also the reduced form. A reduced form should be indistinguishable from a a single sample from the outside, in terms of data required, and the methods on `ErrorIndicators` enable this. The constructor from a `vector<ErrorIndicators>` is a compromise to your request, of allowing for externally collecting indicators before constructing a reduced form.

Using a `GridFunction` for the local indicators has the effect of making all the SEMR loop now have to accept all the crud associated with the `GridFunction` interface, and breaks even further when considering the finite element spaces now defined in spaceoperator etc. will have been destructed by that point, meaning most of the grid functions methods will break due to the dangling pointer. An `ErrorIndicator` needs to stand in isolation without any references, as it's the one thing that'll be departing the `Solve` scope. So using `GridFunction` here is a hard no.